g3doc/design/runtime_dedicate_os_thread.md

```diff
@@ -0,0 +1,151 @@
+syscalls:
+
+-   Task goroutines, which back application threads, are usually blocked on
+    system calls that execute untrusted application code, or (on platforms that
+    do not support synchronous switches to application code, such as ptrace)
+    wait for such execution to complete.
+
+-   The `fdnotifier` package, which serves a similar purpose to the Go runtime's
+    netpoller but sends notifications to types defined in the `waiter` package
+    in order to efficiently support the sentry's implementation of application
+    syscalls such as `poll`, is usually blocked in the `epoll_pwait` syscall.
+
+-   Inbound network packets are processed by "dispatch loop" goroutines that
+    will usually be blocked in the `readv` or `recvmmsg` syscalls (to receive
+    new packets).
+
+Since one task goroutine is needed per application thread, it is common for the
+sentry to be running hundreds or thousands of task goroutines simultaneously,
+each of which requires its own M. This creates an immense scalability burden on
+the Go runtime scheduler.
+
+## Proposal
+
+We propose to add a function, tentatively called `DedicateOSThread`, to the Go
+`runtime` package, documented as follows:
+
+```go
+// DedicateOSThread wires the calling goroutine to its current operating system
+// thread, and exempts it from counting against GOMAXPROCS. The calling
+// goroutine will always execute in that thread, and no other goroutine will
+// execute in it, until the calling goroutine has made as many calls to
+// UndedicateOSThread as to DedicateOSThread. If the calling goroutine exits
+// without unlocking the thread, the thread will be terminated.
+//
+// DedicateOSThread should only be used by long-lived goroutines that usually
+// block due to blocking system calls, rather than interaction with other
+// goroutines.
+func DedicateOSThread()
+```
+
+Mechanically, `DedicateOSThread` implies `LockOSThread` (i.e. it locks the
+invoking G to its M), but additionally locks the invoking M to its P. Ps locked
+by `DedicateOSThread` are not counted against `GOMAXPROCS`; that is, the actual
+number of Ps in the system (`len(runtime.allp)`) is `GOMAXPROCS` plus the number
+of bound Ps (plus some slack to avoid frequent changes to `runtime.allp`).
+Corollaries:
+
+-   If `runtime.ready` observes that a readied G is locked to a M locked to a P,
+    it immediately wakes the locked M instead of waiting for a future call to
+    `runtime.schedule` to select the readied G in `runtime.findrunnable`.
+
+-   `runtime.stoplockedm` and `runtime.reentersyscall` skip the release of
+    locked Ps; the latter also skips sysmon wakeup. `runtime.stoplockedm` and
+    `runtime.exitsyscall` skip re-acquisition of Ps if one is locked.
+
+-   New goroutines created by goroutines with locked Ps are enqueued on the
+    global run queue rather than the invoking P's local run queue.
+
+While gVisor's use case does not strictly require that the association is
```

Comment:
Thinking out loud: what if threads could be unbound prior to blocking in task.Block? Presumably the state of those goroutines do not require a dedicated system thread. When unblocked, they could rebind to the system thread. The main probably with this is churning P's. (E.g. the P will go from "bound" to "unbound" and back to "bound.)

Reply:
More specifically, I think the main problem would be that, since a bound P necessarily doesn't count against GOMAXPROCS, binding a P may force the wakeup of a new M, and unbinding a P may force the M to block until it can obtain an unbound P. (I've now noted this in the "alternatives considered" section.) So this is something we could try, but I don't think it'd be a clear win.

Reply:
This would be a general concern for sure, but in our case I think that we would have plenty of space for a P to swap with an "unbound" P if that was a fast operation. (Since most Ps will not be active for us, presumably.)

I think that in general there are two valid models that we could move towards: a) giving application threads something must closer to system threads and eliminate scheduler churn or b) having a fixed number of system threads that are driving application threads and tame some of the runtime interactions in a different way. Either way, we'd need to fix the findrunnable and indirect wakeup churn.

This proposal is very much (a). There are advantages to (a), but also disadvantages. For example, we'd want to more carefully consider thread bombs, pid limits, etc. (We have this problem today, but it's probably something we want to think about for a proper fix.) I'd love to also think about types of (b) more explicitly and perhaps variants of (a)?

Reply:
> This would be a general concern for sure, but in our case I think that we would have plenty of space for a P to swap with an "unbound" P if that was a fast operation. (Since most Ps will not be active for us, presumably.)

I'm not sure that's the case; while the majority of goroutines will be task goroutines, there are still plenty of other goroutines that may be running (timer goroutines, the kernel CPU clock ticker, networking stuff, etc.)

> having a fixed number of system threads that are driving application threads and tame some of the runtime interactions in a different way

This would be relatively easy to implement (on some platforms), but compared to kernel thread per application thread, I think it's more sensitive to priority inversions due to kernel thread preemption; we'd have to very carefully measure this (latency distribution with CPU antagonists running on the same system).

Reply:
I'd also be concerned about churn, specifically because we definitely have issues with round-trip latency in applications with very short-lived futex (or other) sleeps. Making Task.Block more expensive is the opposite of what we want.

That said, it would hopefully be easy to experiment with.

Reply:
> I'm not sure that's the case; while the majority of goroutines will be task goroutines, there are still plenty of other goroutines that may be running (timer goroutines, the kernel CPU clock ticker, networking stuff, etc.)

We have lots of additional goroutines, but most are active only briefly (and could presumably be supported by a relatively small number of P's). Note that re: networking, all the per-endpoint goroutines are going away at some point.

> This would be relatively easy to implement (on some platforms), but compared to kernel thread per application thread, I think it's more sensitive to priority inversions due to kernel thread preemption; we'd have to very carefully measure this (latency distribution with CPU antagonists running on the same system).

One interesting consequence of the Sentry's design is that I would think priority inversion may exist, but presumably the consequences would be less severe. Given that everything will boil down to futexes at some point, the scheduler should be able to do the right thing with some minor penalty, but there would be no extreme edge cases (e.g. some thread acquiring a spinlock, being preempted). I'd be curious to hear if you had worse-case examples in mind.

> I'd also be concerned about churn, specifically because we definitely have issues with round-trip latency in applications with very short-lived futex (or other) sleeps. Making Task.Block more expensive is the opposite of what we want.

Yes, it couldn't be _much_ more expensive. I think there are parts of the design space that would still keep task.Block relatively cheap (esp. if coupled with core runtime improvements to e.g. reduce the findrunnable cost).